pkg/storage/stores/shipper/downloads/table.go

```diff
@@ -0,0 +1,406 @@
+
+	// keep the files collection locked until all the files are downloaded.
+	table.dbsMtx.Lock()
+	go func() {
+		defer table.dbsMtx.Unlock()
+		defer close(table.ready)
+
+		ctx, cancel := context.WithTimeout(ctx, downloadTimeout)
+		defer cancel()
+
+		// Using background context to avoid cancellation of download when request times out.
+		// We would anyways need the files for serving next requests.
+		if err := table.init(ctx); err != nil {
+			level.Error(util.Logger).Log("msg", "failed to download files", "name", table.name)
+		}
+	}()
+
+	return &table
+}
+
+// init downloads all the db files for the table from object storage.
+// it assumes the locking of mutex is taken care of by the caller.
+func (t *Table) init(ctx context.Context) (err error) {
+	defer func() {
+		status := statusSuccess
+		if err != nil {
+			status = statusFailure
+			t.err = err
+
+			// cleaning up files due to error to avoid returning invalid results.
+			for fileName := range t.dbs {
+				if err := t.cleanupDB(fileName); err != nil {
+					level.Error(util.Logger).Log("msg", "failed to cleanup partially downloaded file", "filename", fileName, "err", err)
+				}
+			}
+		}
+		t.metrics.filesDownloadOperationTotal.WithLabelValues(status).Inc()
+	}()
+
+	startTime := time.Now()
+	totalFilesSize := int64(0)
+
+	objects, _, err := t.storageClient.List(ctx, t.name+"/")
+	if err != nil {
+		return
+	}
+
+	level.Debug(util.Logger).Log("msg", fmt.Sprintf("list of files to download for period %s: %s", t.name, objects))
+
+	folderPath, err := t.folderPathForTable(true)
+	if err != nil {
+		return
+	}
+
+	for _, object := range objects {
+		dbName, err := getDBNameFromObjectKey(object.Key)
+		if err != nil {
+			return err
+		}
+
+		filePath := path.Join(folderPath, dbName)
+		df := downloadedFile{}
+
+		err = t.getFileFromStorage(ctx, object.Key, filePath)
+		if err != nil {
+			return err
+		}
+
+		df.mtime = object.ModifiedAt
+		df.boltdb, err = local.OpenBoltdbFile(filePath)
+		if err != nil {
+			return err
+		}
+
+		var stat os.FileInfo
+		stat, err = os.Stat(filePath)
+		if err != nil {
+			return err
+		}
+
+		totalFilesSize += stat.Size()
+
+		t.dbs[dbName] = &df
+	}
+
+	duration := time.Since(startTime).Seconds()
+	t.metrics.filesDownloadDurationSeconds.add(t.name, duration)
+	t.metrics.filesDownloadSizeBytes.add(t.name, totalFilesSize)
+
+	return
+}
+
+// Closes references to all the dbs.
+func (t *Table) Close() {
+	// stop the initialization if it is still ongoing.
+	t.cancelFunc()
+
+	t.dbsMtx.Lock()
+	defer t.dbsMtx.Unlock()
+
+	for name, db := range t.dbs {
+		if err := db.boltdb.Close(); err != nil {
+			level.Error(util.Logger).Log("msg", fmt.Errorf("failed to close file %s for table %s", name, t.name))
+		}
+	}
+
+	t.dbs = map[string]*downloadedFile{}
+}
+
+// Queries all the dbs for index.
+func (t *Table) Query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	if t.err != nil {
+		return t.err
+	}
+
+	t.dbsMtx.RLock()
```

Comment:
I think the code for waiting initialization should be before this. I know it is the case already but it should included in this function and not in the table manager, because If I use table without tablemanager I can get deadlocked here.

Another option is to make Table not exported in this package.

Reply:
May be easier to just rename Table to table.

Reply:
I have made public methods to give a guarantee that usage of them is concurrency safe. The `Table` is not meant to be used without the table manager. Having public methods(for concurrency guarantees) on a private type would look weird. It would be better to move readiness check here from table manager given we want to support using `Table` without table manager.
What do you think?

Reply:
Either you make `Table` => `table` or you move the readiness check here. Up to you !

Reply:
I have pushed the code. Thanks!